New Top Level Menu Item Tutorial, Tutorial ported from @spicyj's Internal Course

ericnakagawa opened the pull request:
* added top level 'Tutorial' section
* added a 2 images to help illustrate tutorial
* ran through the tutorial and made small edits for clarity

vjeux commented:
Do you mind taking a screenshot of the entire page and displaying it here?

spicyj commented:
Last time I went through this, it felt like the "Storing a History" section was too much typing and moving code around. I was thinking that it might be better to simple have a bullet list

```
* Move #1
* Move #2
* Move #3
```

so that we don't need the whole new {squares, xIsNext, move, when} object and just have an array of squares instead. Not sure if that's best though, open to better ideas. You can watch the internal video at this part and I think you'll agree that it dragged on.

spicyj commented:
We need a couple paragraphs explaining the value of immutable data, more than the brief snippet I currently have. I got like 5 separate questions on it.

ericnakagawa commented:
![screencapture-localhost-4000-react-tutorial-tutorial-html-1475620696450](https://cloud.githubusercontent.com/assets/23874/19095089/a666877e-8a48-11e6-890e-4bc55eb8cb27.png)

Full page capture.

lacker commented:
Nit: All of the nav links have a little "linking to an external site" thingy on them.

gaearon commented:
Do we want to start method names with underscores? We don't do this anywhere else in the docs, and it doesn't seem to be popular in the community either.

gaearon commented on docs/_data/nav_docs.yml:
```diff
@@ -1,9 +1,81 @@
+    href: https://github.com/facebook/react/wiki/Examples
+- title: Guides
+  items:
+  - id: why-react
```
What is the reason for adding these back in? I intentionally removed them from docs and left them in docs-old so that we tell new and old content apart and gradually migrate it.

vjeux replied:
Looking at the history of the commits, it's looks like a bad merge?

ericnakagawa replied:
@gaearon Bad merge, had to back out of it. This PR should reapply the changes you made.

ericnakagawa commented:
@spicyj Added a light section on immutability.

ericnakagawa commented:
@gaearon Removed underscores.

gaearon commented:
Let's get it in and address any nits later? It's getting hard to review with this PR size.

ericnakagawa commented:
@spicyj I simplified Save History to only track squares, xIsNext converted into a function.

ericnakagawa commented:
@lacker Removed external link classes on sidebar by checking for presence of variable in dev_tutorial.yml

ericnakagawa commented:
Updated full page capture:

![screencapture-localhost-4000-react-tutorial-tutorial-html-1475713523946](https://cloud.githubusercontent.com/assets/23874/19136348/cc72e11e-8b20-11e6-8fb3-67a74dbf3df1.png)

spicyj commented:
You mixed up moveNext and moveNumber. Do we need either though? Maybe we can use the index. Though that isn't great for convincing people to key things accurately.

ericnakagawa commented:
@spicyj Resolved variable mixup in docs. Kept moveNumber, removed moveNext. Also removed helper function and calculate next player based on move number when user clicks on history.

spicyj commented:
I meant: can't you infer the move number from the index in array? I'm trying to avoid the {squares, moveNumber} object completely.

lacker commented:
Let's put the "Tutorial" tab second, right after "Docs", just to nudge it up in prominence a bit.

ericnakagawa commented:
@spicyj Removed moveNumber and now using index.

@lacker Moved Tutorial to #2 position on top nav bar.

gaearon commented on docs/tutorial/tutorial.md:
````diff
@@ -0,0 +1,509 @@
+#### Data change without mutation
+```javascript
+var player = {score: 1}
+player = {...player, score: 2} // new object not mutated {score: 2}
````
This is not part of ES6. We should make it clear this syntax is a stage 3 proposal, and show an ES6 alternative.
It might also be confusing why we spread over an object with just one property which we are updating.
